testing/file_test/file_test_base.cpp

```diff
@@ -479,14 +479,84 @@ static auto TransformExpectation(int line_index, llvm::StringRef in)
   if (in.empty()) {
     return Matcher<std::string>{StrEq("")};
   }
-  if (in[0] != ' ') {
+  if (!in.consume_front(" ")) {
     return ErrorBuilder() << "Malformated CHECK line: " << in;
   }
-  std::string str = in.substr(1).str();
+  // Try to avoid an extra string copy if the input is already in a viable form.
+  if (in.find("[[") == llvm::StringRef::npos &&
+      in.find("{{") == llvm::StringRef::npos) {
+    return Matcher<std::string>{StrEq(in)};
+  }
+
+  // We'll have to either insert special sequences or process as a regex. Either
+  // way, we need scratch space so create a copy.
+  std::string str = in.str();
+
+  // First scan the string to expand any keywords and see if we need to process
+  // it as a regex. We don't expand the regex here to simplify escaping below.
+  bool has_regex = false;
+  for (int pos = 0; pos < static_cast<int>(str.size());) {
+    switch (str[pos]) {
+      case '[': {
+        llvm::StringRef line_keyword_cursor = llvm::StringRef(str).substr(pos);
+        if (!line_keyword_cursor.consume_front("[[")) {
+          // Just step over a single square bracket.
+          ++pos;
+          break;
+        }
+
+        static constexpr llvm::StringLiteral LineKeyword = "@LINE";
+        if (!line_keyword_cursor.consume_front(LineKeyword)) {
+          return ErrorBuilder()
+                 << "Unexpected [[, should be {{\\[\\[}} at `"
+                 << line_keyword_cursor.substr(0, 5) << "` in: " << in;
+        }
+
+        // Allow + or - here; consumeInteger handles -.
+        line_keyword_cursor.consume_front("+");
+        int offset;
+        // consumeInteger returns true for errors, not false.
+        if (line_keyword_cursor.consumeInteger(10, offset) ||
+            !line_keyword_cursor.consume_front("]]")) {
+          return ErrorBuilder()
+                 << "Unexpected @LINE offset at `"
+                 << line_keyword_cursor.substr(0, 5) << "` in: " << in;
+        }
+        std::string int_str = llvm::Twine(line_index + offset).str();
+        int remove_len = (line_keyword_cursor.data() - str.data()) - pos;
+        str.replace(pos, remove_len, int_str);
+        pos += int_str.size();
+        break;
+      }
+      case '{': {
+        if (pos + 1 != static_cast<int>(str.size()) && str[pos + 1] != '{') {
+          // Save that we found a regex region, we'll expand it below.
+          has_regex = true;
+          // Do an extra increment to move past the second curly.
+          ++pos;
+        }
+        ++pos;
+        break;
+      }
```

Comment:
Above, you do a `.find("{{"}`. Does it work if you use the result of that for `has_regex`, i.e.:

```
  // Try to avoid an extra string copy if the input is already in a viable form.
  bool has_regex = in.find("{{") != llvm::StringRef::npos;
  if (in.find("[[") == llvm::StringRef::npos && !has_regex) {
    return Matcher<std::string>{StrEq(in)};
  }
```

If that's equivalent, you could remove this case, at which point the loop would only be handling `[[`. That in turn you might be able to handle as:

```
for (int pos = str.find("[["); pos != std::string::npos; pos = str.find("[[", pos))
```

That also has a side-effect of skipping straight past single `[` characters (removing a minor condition), and you could reuse the `in.find("[[")` result above to initialize `pos` too.

In combination, that might also reduce this code enough to address something else I might comment on -- you're adding about 40 lines to this function, taking it to 120-ish lines, which seems a bit high of a line count [for style](https://google.github.io/styleguide/cppguide.html#Write_Short_Functions) right now.

Reply:
Done and extracted a function to keep the size down. It seemed much more readable to extract the regex expansion as that doesn't have lots of error paths or contextual short circuiting, it just processes the string top to bottom. Seemed pretty clean, but lemme know if you'd like anything different. PTAL.